sampling/speculative_sampling.py

```diff
@@ -1,185 +1,210 @@
-                
-                if r < torch.min(torch.tensor([1], device=q.device), p[:, prefix_len + i - 1, j] / q[:, prefix_len + i - 1, j]):
-                    # accept, and update n
-                    n += 1
-                else:
-                    # reject
-                    t = sample(max_fn(p[:, n, :] - q[:, n, :]))
-                    is_all_accept = False
-                    break
-         
-            prefix = x[:, :n + 1]
-            
-            if is_all_accept:
-                t = sample(p[:, -1, :])
-            
-            prefix = torch.cat((prefix, t), dim=1)
-            pbar.update(n - pbar.n)
-
-    return prefix
-
+import torch
+from tqdm import tqdm
+import torch
+
+from sampling.kvcache_model import KVCacheModel
+from sampling.utils import norm_logits, sample, max_fn
+from globals import Decoder
+
+
+@torch.no_grad()
+def speculative_sampling(prefix: torch.Tensor, approx_model: torch.nn.Module, target_model: torch.nn.Module,
+                         max_len: int, gamma: int = 4,
+                         temperature: float = 1, top_k: int = 0, top_p: float = 0, verbose: bool = False,
+                         random_seed: int = None) -> torch.Tensor:
+    """
+    Google version Speculative Sampling.
+    https://arxiv.org/pdf/2211.17192.pdf
+
+    Adapted with KV Cache Optimization.
+
+    Args:
+        x (torch.Tensor): input sequence, (batch, prefix_seqlen), Note that the batch dim is always 1 now.
+        approx_model (torch.nn.Module): approx model, the small one
+        target_model (torch.nn.Module): target model, the large one
+        max_len (int): the max overall generated tokens number.
+        gamma (int): $\gamma$, the token number small model guesses.
+        temperature (float, optional): Defaults to 1.
+        top_k (int, optional): Defaults to 0.
+        top_p (float, optional): Defaults to 0.
+
+    Returns:
+        torch.Tensor: generated tokens (batch, target_seqlen)
+    """
+    seq_len = prefix.shape[1]
+    T = seq_len + max_len
+
+    assert prefix.shape[0] == 1, "input batch size must be 1"
+
+    assert approx_model.device == target_model.device
+
+    device = target_model.device
+
+    approx_model_cache = KVCacheModel(approx_model, temperature, top_k, top_p)
+    target_model_cache = KVCacheModel(target_model, temperature, top_k, top_p)
+
+    resample_count = 0
+    target_sample_count = 0
+    accepted_count = 0
+
+    while prefix.shape[1] < T:
+        # q = M_q[prefix + x_0, x_1, .., x_(gamma-2)]
+        prefix_len = prefix.shape[1]
+
+        x = approx_model_cache.generate(prefix, gamma)
+        _ = target_model_cache.generate(x, 1)
+
+        n = prefix_len + gamma - 1
+
+        for i in range(gamma):
+            if random_seed:
+                torch.manual_seed(random_seed)
+            r = torch.rand(1, device=device)
+            j = x[:, prefix_len + i]
+
+            if r > (target_model_cache._prob_history[:, prefix_len + i - 1, j]) / (

```

Comment:
why not also change to `target_model_cache._prob_history.p[i, -1, j]`

Reply:
I made no change on this function. It's as it was.